### Import neccessary libraries here

In [1]:
import pandas as pd
import numpy as np
import seaborn as sns
import matplotlib as plt

# Let's load our csv in...
### Because it is > 100 MB, we have to store it on our local machine instead of github, you'll have to comment out my directory and put in your own

In [2]:
ct = pd.read_csv("C:/Users/conor/Desktop/city_temperature/city_temperature.csv")
ct.head(5)

C:\Users\conor\anaconda3\lib\site-packages\IPython\core\interactiveshell.py:3437: DtypeWarning: Columns (2) have mixed types.Specify dtype option on import or set low_memory=False.
  exec(code_obj, self.user_global_ns, self.user_ns)


Region  Country State     City  Month  Day  Year  AvgTemperature
0  Africa  Algeria   NaN  Algiers      1    1  1995            64.2
1  Africa  Algeria   NaN  Algiers      1    2  1995            49.4
2  Africa  Algeria   NaN  Algiers      1    3  1995            48.8
3  Africa  Algeria   NaN  Algiers      1    4  1995            46.4
4  Africa  Algeria   NaN  Algiers      1    5  1995            47.9

In [3]:
ct.tail(5)

Region Country                   State                  City  \
2906322  North America      US  Additional Territories  San Juan Puerto Rico   
2906323  North America      US  Additional Territories  San Juan Puerto Rico   
2906324  North America      US  Additional Territories  San Juan Puerto Rico   
2906325  North America      US  Additional Territories  San Juan Puerto Rico   
2906326  North America      US  Additional Territories  San Juan Puerto Rico   

         Month  Day  Year  AvgTemperature  
2906322      7   27  2013            82.4  
2906323      7   28  2013            81.6  
2906324      7   29  2013            84.2  
2906325      7   30  2013            83.8  
2906326      7   31  2013            83.6

### First, lets see how many rows and columns we have.

In [4]:
#.shape is an attribute of DataFrames that returns (#rows, #columns)
ct.shape

(2906327, 8)

### Next, lets check for any null values in the rows.


In [5]:
#check if any cell in the column is a null value
ct.isna().any()

Region            False
Country           False
State              True
City              False
Month             False
Day               False
Year              False
AvgTemperature    False
dtype: bool

### It looks like we have NaN values in the 'State' column. We known how many rows exist in the data set so let's find out how many rows in 'State' are NaN

In [6]:
#sum the null values in each column
ct.isna().sum()

Region                  0
Country                 0
State             1450990
City                    0
Month                   0
Day                     0
Year                    0
AvgTemperature          0
dtype: int64

### Let's look at those rows where the State column is not null

In [7]:
#subset our ct datset by where the state column is not null
ct[ct['State'].notna()]

Region Country                   State                  City  \
1450990  North America      US                 Alabama            Birmingham   
1450991  North America      US                 Alabama            Birmingham   
1450992  North America      US                 Alabama            Birmingham   
1450993  North America      US                 Alabama            Birmingham   
1450994  North America      US                 Alabama            Birmingham   
...                ...     ...                     ...                   ...   
2906322  North America      US  Additional Territories  San Juan Puerto Rico   
2906323  North America      US  Additional Territories  San Juan Puerto Rico   
2906324  North America      US  Additional Territories  San Juan Puerto Rico   
2906325  North America      US  Additional Territories  San Juan Puerto Rico   
2906326  North America      US  Additional Territories  San Juan Puerto Rico   

         Month  Day  Year  AvgTemperature  
1450990      1    1  1995            50.7  
1450991      1    2  1995            37.2  
1450992      1    3  1995            33.2  
1450993      1    4  1995            33.3  
1450994      1    5  1995            26.4  
...        ...  ...   ...             ...  
2906322      7   27  2013            82.4  
2906323      7   28  2013            81.6  
2906324      7   29  2013            84.2  
2906325      7   30  2013            83.8  
2906326      7   31  2013            83.6  

[1455337 rows x 8 columns]

### Hmm it looks like both the head and tail of our subset are both North America... lets see if there are any areas where the region isn't North America and the State is not null

In [8]:
#subset for two conditions now (we need to enclose each condition in parentheses in order to do this)
ct[(ct['Region'] != 'North America') & (ct['State'].notna())]

Empty DataFrame
Columns: [Region, Country, State, City, Month, Day, Year, AvgTemperature]
Index: []

### Next, let's make sure out Months, Days, and Years are all valid

In [9]:
#subset the dataset by rows where the month is between 1 and 12, then lets just return shape to make sure the # of rows is
#same as our original shape
ct[ct['Month'].isin(range(1,13))].shape

(2906327, 8)

In [10]:
#Let's do the same for days, making sure they are between 1 and 31
ct[ct['Day'].isin(range(1,32))].shape

(2906319, 8)

### Hmm... there are 8 rows where the day is not between 1 and 31. Let's take a closer look at those

In [11]:
#by placing a tilde (~) before a condition, it reverses the condition and only selects where the condition is false 
ct[~ct['Day'].isin(range(1,32))]

Region        Country State         City  \
82774                               Africa         Guinea   NaN      Conakry   
85697                               Africa         Guinea   NaN      Conakry   
92041                               Africa  Guinea-Bissau   NaN       Bissau   
146077                              Africa         Malawi   NaN     Lilongwe   
177862                              Africa        Nigeria   NaN        Lagos   
241159                              Africa         Uganda   NaN      Kampala   
1209901                      North America         Mexico   NaN  Guadalajara   
1333910  South/Central America & Carribean           Cuba   NaN       Havana   

         Month  Day  Year  AvgTemperature  
82774        3    0  2008           -99.0  
85697        3    0  2016           -99.0  
92041        3    0  2008           -99.0  
146077       3    0  2012           -99.0  
177862       3    0  2008           -99.0  
241159       3    0  2012           -99.0  
1209901      3    0  2012           -99.0  
1333910      3    0  2008           -99.0

### Interesting. These days are equal to 0 and the Average Temperature is -99.0! We will revisit this after finishing our date time conditions.

In [12]:
#lets sort our dataframe by hear, not specifying tail or head so that we get a snapshot of the earliest and latest years
ct.sort_values('Year')

Region  Country         State         City  Month  Day  Year  \
743771          Europe  Germany           NaN      Hamburg     12   12   200   
845936          Europe   Norway           NaN         Oslo     12   28   200   
845935          Europe   Norway           NaN         Oslo     12   27   200   
845934          Europe   Norway           NaN         Oslo     12   26   200   
845933          Europe   Norway           NaN         Oslo     12   25   200   
...                ...      ...           ...          ...    ...  ...   ...   
2593753  North America       US  South Dakota  Sioux Falls      4    3  2020   
2593752  North America       US  South Dakota  Sioux Falls      4    2  2020   
2593751  North America       US  South Dakota  Sioux Falls      4    1  2020   
794360          Europe  Ireland           NaN       Dublin      1   10  2020   
2476788  North America       US        Oregon      Medford      2   29  2020   

         AvgTemperature  
743771            -99.0  
845936            -99.0  
845935            -99.0  
845934            -99.0  
845933            -99.0  
...                 ...  
2593753            23.5  
2593752            44.3  
2593751            52.4  
794360             41.3  
2476788            47.8  

[2906327 rows x 8 columns]

In [13]:
#We saw that the earliest year was 200... and mysteriously the average temperature for those was also -99.0!
ct[ct['Year'].isin(range(1995,2021))].shape

(2905887, 8)

In [14]:
#So there are lots of values missing now, lets take a look at those
ct[~ct['Year'].isin(range(1995,2021))]

Region   Country State         City  Month  Day  Year  \
58178           Africa  Ethiopia   NaN  Addis Ababa     12    3   201   
58179           Africa  Ethiopia   NaN  Addis Ababa     12    4   201   
58180           Africa  Ethiopia   NaN  Addis Ababa     12    5   201   
58181           Africa  Ethiopia   NaN  Addis Ababa     12    6   201   
58182           Africa  Ethiopia   NaN  Addis Ababa     12    7   201   
...                ...       ...   ...          ...    ...  ...   ...   
1212427  North America    Mexico   NaN  Guadalajara     12   27   201   
1212428  North America    Mexico   NaN  Guadalajara     12   28   201   
1212429  North America    Mexico   NaN  Guadalajara     12   29   201   
1212430  North America    Mexico   NaN  Guadalajara     12   30   201   
1212431  North America    Mexico   NaN  Guadalajara     12   31   201   

         AvgTemperature  
58178             -99.0  
58179             -99.0  
58180             -99.0  
58181             -99.0  
58182             -99.0  
...                 ...  
1212427           -99.0  
1212428           -99.0  
1212429           -99.0  
1212430           -99.0  
1212431           -99.0  

[440 rows x 8 columns]

In [15]:
#Again, lots of -99.0 average temperatures, with lots in places that are notoriously hot. Lets see if there are any places
#where the year is out of range and the temperature isn't -99.0.
ct[(~ct['Year'].isin(range(1995,2021))) & (ct['AvgTemperature'] != -99.0)]

Empty DataFrame
Columns: [Region, Country, State, City, Month, Day, Year, AvgTemperature]
Index: []

### So there are 440 rows where the year isn't in our desired range, and surpisingly the average temperature for all of those values is -99.0. Let's apply the same logic we did to our date time columns to our AvgTemperature columns to make sure those values are all logical.

In [16]:
ct.sort_values('AvgTemperature')

Region        Country State        City  \
1377340  South/Central America & Carribean         Guyana   NaN  Georgetown   
88779                               Africa  Guinea-Bissau   NaN      Bissau   
88778                               Africa  Guinea-Bissau   NaN      Bissau   
88777                               Africa  Guinea-Bissau   NaN      Bissau   
88776                               Africa  Guinea-Bissau   NaN      Bissau   
...                                    ...            ...   ...         ...   
1034963                        Middle East         Kuwait   NaN      Kuwait   
1036756                        Middle East         Kuwait   NaN      Kuwait   
1032390                        Middle East         Kuwait   NaN      Kuwait   
1036042                        Middle East         Kuwait   NaN      Kuwait   
1034962                        Middle East         Kuwait   NaN      Kuwait   

         Month  Day  Year  AvgTemperature  
1377340      8   21  2005           -99.0  
88779        3   27  1999           -99.0  
88778        3   26  1999           -99.0  
88777        3   25  1999           -99.0  
88776        3   24  1999           -99.0  
...        ...  ...   ...             ...  
1034963      8    2  2012           109.9  
1036756      6   29  2017           109.9  
1032390      7   17  2005           109.9  
1036042      7   17  2015           109.9  
1034962      8    1  2012           110.0  

[2906327 rows x 8 columns]

### 110.0 Fahrenheit is a believable temperature but -99.0 Fahrenheit in August in Guyana? I'm not sure thats accurate. Let's now grab unique temperature values

In [17]:
#We create a set of values using .unique() and then sort it. The default sorting is ascending
sorted(ct['AvgTemperature'].unique())[:5]

[-99.0, -50.0, -49.1, -47.7, -46.9]

### Theres a 49.0 degree gap between the lowest number (-99.0) and the second lowest (-50.0). Let's check -50.0 to make sure it seems like a legitimate reading.

In [18]:
#Use .loc (label based lookup) to see the row where AvgTemperature was -50.0
ct.loc[ct['AvgTemperature'] == -50.0]

Region Country   State       City  Month  Day  Year  \
1499140  North America      US  Alaska  Fairbanks     12   31  1999   

         AvgTemperature  
1499140           -50.0

### Our true lowest record temperature was on December 31st in Fairbanks, Alaska. Sounds good to me!

### So we know there are some days in which day = 0, year = 200 or 201, and AvgTemp = -99.0. Let's remove those from out dataset.

In [19]:
ct_fltrd = ct[(ct['Day'].isin(range(1,32))) & (ct['Year'].isin(range(1995,2021))) & (ct['AvgTemperature'] != -99.0)]
ct_fltrd

Region  Country                   State                  City  \
0               Africa  Algeria                     NaN               Algiers   
1               Africa  Algeria                     NaN               Algiers   
2               Africa  Algeria                     NaN               Algiers   
3               Africa  Algeria                     NaN               Algiers   
4               Africa  Algeria                     NaN               Algiers   
...                ...      ...                     ...                   ...   
2906322  North America       US  Additional Territories  San Juan Puerto Rico   
2906323  North America       US  Additional Territories  San Juan Puerto Rico   
2906324  North America       US  Additional Territories  San Juan Puerto Rico   
2906325  North America       US  Additional Territories  San Juan Puerto Rico   
2906326  North America       US  Additional Territories  San Juan Puerto Rico   

         Month  Day  Year  AvgTemperature  
0            1    1  1995            64.2  
1            1    2  1995            49.4  
2            1    3  1995            48.8  
3            1    4  1995            46.4  
4            1    5  1995            47.9  
...        ...  ...   ...             ...  
2906322      7   27  2013            82.4  
2906323      7   28  2013            81.6  
2906324      7   29  2013            84.2  
2906325      7   30  2013            83.8  
2906326      7   31  2013            83.6  

[2826655 rows x 8 columns]

### Lets check all of the scenarios in which our data is out of range or values are incorrect to make sure we didn't lose any more/less rows than we expected to

In [34]:
#Day wrong, Year and Temp in range
print(ct[(~ct['Day'].isin(range(1,32))) & (ct['Year'].isin(range(1995,2021))) & (ct['AvgTemperature'] != -99.0)].shape)
#Year wrong, Day and Temp in range
print(ct[(ct['Day'].isin(range(1,32))) & (~ct['Year'].isin(range(1995,2021))) & (ct['AvgTemperature'] != -99.0)].shape)
#AvgTemp wrong, Day and year in range
print(ct[(ct['Day'].isin(range(1,32))) & (ct['Year'].isin(range(1995,2021))) & (ct['AvgTemperature'] == -99.0)].shape)
#Day and AvgTemp wrong, Year in range
print(ct[~(ct['Day'].isin(range(1,32))) & (ct['Year'].isin(range(1995,2021))) & (ct['AvgTemperature'] == -99.0)].shape)
#Day and Year wrong, Temp in range
print(ct[~(ct['Day'].isin(range(1,32))) & (~ct['Year'].isin(range(1995,2021))) & (ct['AvgTemperature'] != -99.0)].shape)
#Year and Avg Temp wrong, Day in range
print(ct[(ct['Day'].isin(range(1,32))) & (~ct['Year'].isin(range(1995,2021))) & (ct['AvgTemperature'] == -99.0)].shape)

(0, 8)
(0, 8)
(79224, 8)
(8, 8)
(0, 8)
(440, 8)


In [37]:
#Let's sum the rows from the invalid cases
print((79224+8+440))

79672


In [38]:
#Lets use shape to determine how many rows we lost after we filtered the data
ct.shape[0]-ct_fltrd.shape[0]

79672

### Perfect. Our filtered data is the exact size that we expected it to be. We can not begin working with it!

In [39]:
ct_fltrd.to_csv(r'C:/Users/conor/Desktop/city_temperature/city_temperature_filtered.csv')